File: packages/sdk-flutter/test/endpoint_validator_test.dart
import 'package:eodin_sdk/src/internal/endpoint_validator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('validateEndpoint — accepted', () {
    test('https://... passes', () {
      expect(() => validateEndpoint('https://api.eodin.app/api/v1'), returnsNormally);
      expect(() => validateEndpoint('https://api-staging.eodin.app/api/v1'), returnsNormally);
      expect(() => validateEndpoint('https://api.eodin.app:443/api/v1'), returnsNormally);
    });

    test('http://localhost / 127.0.0.1 passes', () {
      expect(() => validateEndpoint('http://localhost:3005/api/v1'), returnsNormally);
      expect(() => validateEndpoint('http://127.0.0.1:3005/api/v1'), returnsNormally);
    });

    test('http://10.0.2.2 passes in debug build (test runner = debug)', () {
      // flutter_test runs in debug mode (kDebugMode = true).
      expect(() => validateEndpoint('http://10.0.2.2:3005/api/v1'), returnsNormally);
    });

    test('case-insensitive scheme: HTTPS, HtTp localhost', () {
      expect(() => validateEndpoint('HTTPS://api.eodin.app/api/v1'), returnsNormally);
      expect(() => validateEndpoint('Http://localhost:3005/api/v1'), returnsNormally);
    });

    test('whitespace trimmed', () {
      expect(() => validateEndpoint('  https://api.eodin.app/api/v1  '), returnsNormally);
    });
  });

  group('validateEndpoint — rejected', () {
    test('plain http on non-loopback throws ArgumentError', () {
      expect(() => validateEndpoint('http://api.eodin.app/api/v1'),
          throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('http://attacker.example.com/collect'),
          throwsA(isA<ArgumentError>()));
    });

    test('empty / whitespace-only throws ArgumentError', () {
      expect(() => validateEndpoint(''), throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('   '), throwsA(isA<ArgumentError>()));
    });

    test('non-URL string throws ArgumentError', () {
      expect(() => validateEndpoint('not-a-url'), throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('://no-scheme'), throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('https://'), throwsA(isA<ArgumentError>()));
    });

    test('unsupported scheme (ftp, ws, file) throws ArgumentError', () {
      expect(() => validateEndpoint('ftp://api.eodin.app'), throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('ws://api.eodin.app'), throwsA(isA<ArgumentError>()));
      expect(() => validateEndpoint('file:///etc/passwd'), throwsA(isA<ArgumentError>()));
    });

    test('confusable host like api.eodin.app.attacker.com is HTTPS so passes (host whitelist 보류 — open-issues §4.6)', () {
      // 본 테스트는 *현재 동작* 을 명시적으로 표현 — host 화이트리스트는 v2.x 보류 결정
      expect(() => validateEndpoint('https://api.eodin.app.attacker.com'),
          returnsNormally);
    });
  });
}

File: packages/sdk-android/src/test/java/app/eodin/internal/EndpointValidatorTest.kt
package app.eodin.internal

import org.junit.Assert.assertThrows
import org.junit.Test

class EndpointValidatorTest {
    // ---- Accepted ----

    @Test
    fun `https passes`() {
        EndpointValidator.validate("https://api.eodin.app/api/v1")
        EndpointValidator.validate("https://api-staging.eodin.app/api/v1")
        EndpointValidator.validate("https://api.eodin.app:443/api/v1")
    }

    @Test
    fun `http loopback passes`() {
        EndpointValidator.validate("http://localhost:3005/api/v1")
        EndpointValidator.validate("http://127.0.0.1:3005/api/v1")
    }

    @Test
    fun `10_0_2_2 passes in debug build`() {
        // BuildConfig stub has DEBUG = true; in Gradle release build it would
        // be regenerated to false and this would throw — but unit tests run
        // against the debug stub, so we only verify the debug branch here.
        EndpointValidator.validate("http://10.0.2.2:3005/api/v1")
    }

    @Test
    fun `case-insensitive scheme`() {
        EndpointValidator.validate("HTTPS://api.eodin.app/api/v1")
        EndpointValidator.validate("Http://localhost:3005/api/v1")
    }

    @Test
    fun `whitespace trimmed`() {
        EndpointValidator.validate("  https://api.eodin.app/api/v1  ")
    }

    // ---- Rejected ----

    @Test
    fun `plain http on non-loopback throws`() {
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("http://api.eodin.app/api/v1")
        }
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("http://attacker.example.com/collect")
        }
    }

    @Test
    fun `empty throws`() {
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("")
        }
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("   ")
        }
    }

    @Test
    fun `non-URL throws`() {
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("not-a-url")
        }
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("https://")
        }
    }

    @Test
    fun `unsupported scheme throws`() {
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("ftp://api.eodin.app")
        }
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("ws://api.eodin.app")
        }
        assertThrows(IllegalArgumentException::class.java) {
            EndpointValidator.validate("file:///etc/passwd")
        }
    }

    @Test
    fun `confusable host passes — host whitelist deferred to open-issues 4_6`() {
        EndpointValidator.validate("https://api.eodin.app.attacker.com")
    }
}
